framework/sandstone_ssl_rand.cpp

```diff
@@ -0,0 +1,160 @@
+ * https://www.openssl.org/docs/man3.0/man7/EVP_RAND-CTR-DRBG.html
+ */
+static int sandstone_rand_get_ctx_params(void *ctx, OSSL_PARAM params[])
+{
+    OSSL_PARAM *p;
+
+    p = s_OSSL_PARAM_locate(params, OSSL_RAND_PARAM_STATE);
+    if (p != NULL && !s_OSSL_PARAM_set_int(p, *(int *)ctx))
+        return 0;
+
+    p = s_OSSL_PARAM_locate(params, OSSL_RAND_PARAM_STRENGTH);
+    if (p != NULL && !s_OSSL_PARAM_set_int(p, 500))
+        return 0;
+
+    p = s_OSSL_PARAM_locate(params, OSSL_RAND_PARAM_MAX_REQUEST);
+    if (p != NULL && !s_OSSL_PARAM_set_size_t(p, INT_MAX))
+        return 0;
+
+    return 1;
+}
+
+
+/* Random functions
+ * https://www.openssl.org/docs/man3.0/man7/provider-rand.html
+ */
+static const OSSL_DISPATCH sandstone_rand_functions[] = {
+    {OSSL_FUNC_RAND_NEWCTX, (void (*)(void))sandstone_rand_newctx},
+    {OSSL_FUNC_RAND_FREECTX, (void (*)(void))sandstone_rand_freectx},
+    {OSSL_FUNC_RAND_INSTANTIATE, (void (*)(void))sandstone_rand_instantiate},
+    {OSSL_FUNC_RAND_UNINSTANTIATE, (void (*)(void))sandstone_rand_uninstantiate},
+    {OSSL_FUNC_RAND_GENERATE, (void (*)(void))sandstone_rand_generate},
+    {OSSL_FUNC_RAND_RESEED, (void (*)(void))sandstone_rand_reseed},
+    {OSSL_FUNC_RAND_ENABLE_LOCKING, (void (*)(void))sandstone_rand_enable_locking},
+    {OSSL_FUNC_RAND_GETTABLE_CTX_PARAMS, (void(*)(void))sandstone_rand_gettable_ctx_params },
+    {OSSL_FUNC_RAND_GET_CTX_PARAMS, (void(*)(void))sandstone_rand_get_ctx_params },
+    {0, NULL}
+};
+
+static const OSSL_ALGORITHM sandstone_rand_op[] = {
+    {"sand-rand", "provider=sandstone-rand", sandstone_rand_functions},
+    {NULL, NULL, NULL}
+};
+
+static const OSSL_ALGORITHM *sandstone_rand_query(const OSSL_PROVIDER *prov,
+    int operation_id, int *no_cache)
+{
+    // Provider will only offer random
+    switch (operation_id) {
+        case OSSL_OP_RAND:
+            return sandstone_rand_op;
+    }
+    return NULL;
+}
+
+/* Base functions offered by the provider */
+static const OSSL_DISPATCH sandstone_rand_method[] = {
+    {OSSL_FUNC_PROVIDER_TEARDOWN, (void (*)(void))OSSL_LIB_CTX_free},
+    {OSSL_FUNC_PROVIDER_QUERY_OPERATION, (void (*)(void))sandstone_rand_query},
+    {0, NULL }
+};
+
+/* Provider's initialization */
+static int sandstone_ssl_rand_provider_init(const OSSL_CORE_HANDLE *handle,
+    const OSSL_DISPATCH *in, const OSSL_DISPATCH **out,
+    void **provctx)
+{
+    // Create provider context
+    *provctx = s_OSSL_LIB_CTX_new();
+    if (*provctx == NULL)
+         return 0;
+
+    *out = sandstone_rand_method;
+    return 1;
+}
+
+// TODO: Add clean up funcitons
+void sandstone_ssl_rand_init()
+{
+    if (s_OSSL_PROVIDER_add_builtin(NULL, "sandstone-rand", sandstone_ssl_rand_provider_init) != 1
+            || s_RAND_set_DRBG_type(NULL, "sand-rand", NULL, NULL, NULL) != 1
+            || (s_prov = s_OSSL_PROVIDER_try_load(NULL, "sandstone-rand", 1)) == NULL )
+        printf("WARNING: Cannot load sandstone random provider");
```

Comment:
`fprintf` to `stderr`; please start the warning with "%s" / `program_invocation_name`.